pyop2/compilation.py

```diff
@@ -112,26 +138,38 @@ def get_so(self, src, extension):
         # atomically (avoiding races).
         tmpname = os.path.join(cachedir, "%s_p%d.so.tmp" % (basename, pid))
 
+        if MPI.version >= 3:
+            # Assume that the library supports flashy MPI3 features
+            newcomm = self.comm.Split_type(MPI.COMM_TYPE_SHARED)
+            debug("Using MPI3 on rank ", newcomm.rank)
+            if newcomm.rank == 0:
+                singleton = True
+        else:
+            newcomm = self.comm
+            debug("Not using MPI3 on %d" % newcomm.rank)
+            singleton = self.single_out_a_rank(newcomm)
```

Comment:
I would prefer if the logic was:

```
newcomm = split_comm(comm)
```

where `split_comm` either splits based on `MPI_COMM_TYPE_SHARED`, or else on the file-based system.